sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,374 @@
+                                              uint32_t mask) {
+  if (gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  mmio_region_nonatomic_clear_mask32(gpio->base_addr,
+                                     GPIO_INTR_ENABLE_REG_OFFSET, mask, 0);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_irq_trigger_masked_disable(const dif_gpio_t *gpio,
+                                                      uint32_t mask) {
+  if (gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  // Disable all interrupt triggers for the given mask
+  mmio_region_nonatomic_clear_mask32(
+      gpio->base_addr, GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, mask, 0);
+  mmio_region_nonatomic_clear_mask32(
+      gpio->base_addr, GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, mask, 0);
+  mmio_region_nonatomic_clear_mask32(
+      gpio->base_addr, GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, mask, 0);
+  mmio_region_nonatomic_clear_mask32(
+      gpio->base_addr, GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, mask, 0);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_irq_trigger_masked_config(
+    const dif_gpio_t *gpio, uint32_t mask, dif_gpio_irq_t config) {
+  if (gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  // Disable all interrupt triggers for the given mask
+  dif_gpio_irq_trigger_masked_disable(gpio, mask);
+
+  switch (config) {
+    case kDifGpioIrqEdgeRising:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, mask, 0);
+      break;
+    case kDifGpioIrqEdgeFalling:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, mask, 0);
+      break;
+    case kDifGpioIrqLevelLow:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, mask, 0);
+      break;
+    case kDifGpioIrqLevelHigh:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, mask, 0);
+      break;
+    case kDifGpioIrqEdgeRisingFalling:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, mask, 0);
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, mask, 0);
+      break;
+    case kDifGpioIrqEdgeRisingLevelLow:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, mask, 0);
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, mask, 0);
+      break;
+    case kDifGpioIrqEdgeFallingLevelHigh:
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, mask, 0);
+      mmio_region_nonatomic_set_mask32(
+          gpio->base_addr, GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, mask, 0);
+      break;
+    default:
+      return kDifGpioResultInvalidArgument;
```

Comment:
Please can you change this to `return kDifGpioResultGeneralError` or similar? I prefer the "Invalid Argument" return code to also signal that the function did not have any side effects. The reason to keep the `default` is not just argument checking, it's also in case the value you're switching on has been changed between entering the function and executing the switch.

Reply:
Good point, thanks! Done. I also verified that this is the only place where an error condition can have a side effect.